### ISSUE TITLE
HttpClient docs : Type-checking not working with error handling exemple

### ISSUE THREAD

**Issue by ThomasKientz** [angular/angular#22818]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
https://angular.io/guide/http#getting-error-details

The doc provides this exemple :

```
getConfig() {
  return this.http.get<Config>(this.configUrl)
    .pipe(
      catchError(this.handleError)
    );
}
```

getConfig() returns an `Observable<any>`

## Expected behavior

getConfig() returns an `Observable<Config>`

## Minimal reproduction of the problem with instructions

https://stackblitz.com/edit/angular-8xgmuw

Check comments in `hello.component.ts`

My solution : 

```
return this.http.get<Config>('configUrl')
      .pipe(
        catchError<Config, Config>(this.handleError)
      );
```
## What is the motivation / use case for changing the behavior?

Improve the doc

## Environment

<pre><code>
Angular version: 5.2.8

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

**Comment by dawidgarus** [comment]
From my experience, if You want `catch` to return observable of the same type, the best way is to make handleError a generic function: `handleError<T>(error: HttpErrorResponse, caught: Observable<T>): Observable<T>`.
`catch` and `catchError` can then properly infer return type.

**Comment by leblancdavid** [comment]
I don't believe it only happens with the `catchError`. It seems that `http.get<T>(...)` (or other calls) always returns `Observable<Object>` regardless of the `<T>`. I've updated the example you provided to show this:

[https://stackblitz.com/edit/angular-eesfrt](https://stackblitz.com/edit/angular-eesfrt)

When I call `getConfig()` on the `configService`, the `data` returned is of type `Object`. This code: 
```
this.configService.getConfig().subscribe(data => {
      //getConfig() should return Observable<Config>
      //therefore, data should be of type <Config>, but it's not!
      //Doing a deep copy also doesn't solve this problem
      this.config = { ...data };
      
      //neither does typecasting
      //this.config = <Config>data;

      //this is the only solution:
      //uncomment this and obviously it will work
      //this.config = new Config(data.uuid);
      
      this.isAConfigType = this.config instanceof Config;
```

I don't know if this is a bug or not but sure feels like one! Am I missing something?

**Comment by ThomasKientz** [comment]
@leblancdavid Why would you do `http.get<T>(...)` ? If you don't have type, you can use <any> or create an interface and specify it : `http.get<YourInterface>(...)`. 

The goal of doing :` handleError<T>(error: HttpErrorResponse, caught: Observable<T>): Observable<T>` is to keep the type provided with `http.get<MyType>(...)` in the response.

**Comment by leblancdavid** [comment]
@ThomasKientz I'm a bit confused. I would expect when doing `http.get<Config>(...)` the result I would get when calling `subscribe( result => { ... })` would be a `Config` wouldn't it? What I am seeing is that it is actually returning an `Object`. So far the only way I found to ensure that I actually get an instance of `Config` is to use `map()` as follows:

```
@Injectable()
export class ConfigService {
  constructor(private http: HttpClient) { }

  getConfig(): Observable<Config> {
    return this.http.get<Config>('https://httpbin.org/uuid')
     .map(result => {
      return new Config(result.id, .... , );
   });
  }
}
```

If you run the stackblitz I provided above you will see it.

**Comment by ThomasKientz** [comment]
@leblancdavid For what I understand, it looks like expected : Typescript recognizes it as a `Config` type. But, `instanceof` is a Javascript method, it has nothing to do with Typescript.

**Comment by leblancdavid** [comment]
@ThomasKientz I see, that makes sense. However I noticed this issue when I tried to call on the method on the class I get back from the GET. For example, if my `Config` class had a method called `foo() {...}` and I tried to do:

```
this.configService.getConfig().subscribe(config => {
     config.foo(); //this throws "config.foo is not a function" ERROR
```

Am I missing something? I updated the [example](https://stackblitz.com/edit/angular-eesfrt):

**Comment by ThomasKientz** [comment]
@leblancdavid this could help you https://medium.com/front-end-hacking/typescript-class-vs-interface-99c0ae1c2136?source=linkShare-7dd0d4d89120-1523000250
